### ISSUE TITLE
Kodi compile fails on pine64

### ISSUE THREAD

**Issue by streppuiu** [OpenELEC/OpenELEC.tv#4994]
Hi, I am trying to compile openelec 6.0.0 and 7.0.0 for raspberry pi on a pine64 with debian-mate-jessie-20160701-lenny.raposo-longsleep-pine64-8GB os on it.
I am receiving the following error when it reaches kodi, on both versions (7 and 8):

configure:20190: error: unsupported native build platform: aarch64-unknown-linux-gnu.

Here is the config.log file I found in the build/kodi folder:
http://openelec.tv/media/kunena/attachments/27642/config.log

Is there some limitation for the platforms I can crosscompile openelec on, or is this some other issue?
Is there something I can do to solve this problem other than using another hardware to build openelec ?
Thanks

**Comment by streppuiu** [comment]
Well, I also like to laugh but I don't usually do it at people that want to learn. That is a sign of stupidity actually as far as I know. So please, give me an answer here to show us that the one who's stupid is not you but me :) . Thank you.

**Comment by chewitt** [comment]
OpenELEC's developer uses Fedora on an Intel x86-64 box and Ubuntu LTS releases are normally okay for cross-compiling images. Debian on low-spec hardware isn't a promising combination :)

**Comment by streppuiu** [comment]
Thank you chewitt, finally an answer, I almost feel stupid now but not quite :) .

I am aware that on a pc+ubuntu it works fine, I compiled oe5 and 6 and I also created a couple of addons for oe/RPi which compiled just fine on that. The pc I used had a pII@700mhz cpu and 128mb of RAM and took quite some space on my frige.

Pine64 has a quad-core 1200mhz cpu and 1gb of RAM and it takes about 15cm2 of space.

Now you tell me which one of the two is a low-spec hw computer. And even if it is pine+deb, I still fail to understand why it is not a promising combination for cross-compiling (wfich is simple code compilation after all). I am not speed I am after.

So, is this error message a limitation of the oe version of Kodi or is it something else? And if it's something else then how to fix it? Thank you.

**Comment by chewitt** [comment]
1200Mhz is fast for ARM but MHz != bogomips and it's still a low-spec CPU that I doubt anyone has ever bothered to compile test on (because an image is large an most people do care about compile times). I'm not going to debate the OE build system further though as I've not paid any attention to it since 13th March. I was just responding to point out the folly of your choices. It might/could work, but you are in uncharted waters.

**Comment by streppuiu** [comment]
I understand perfectly what you say and I really appreciate the time you are taking to answer.

I also understand that I am on my own in this (which is something I like actually :) ).

About the folly :) , I will tell you that I am only an occasional builder, and thus build time is not important for me. I can wait a few days to have something so big as oe compiled, smaller things compile much faster anyway.

I assume that even if I am alone in this, I am entitled to ask for help. And where should I do that if not on the oe dev forum?

So am I to understand from your answer that the OE build system prevents me from running the compilation on that machine because it would take too long to do it? You won't have to debate anything here, a simple yes or no to this question will help me (if you are sure about it). 

If it's not that then what is it? Maybe you or someone else encountered something similar already and could tell me how to solve it.

Thank you.

**Comment by CvH** [comment]
@streppuiu to avoid the problems with debian as base os you could run the compile process in a [dedicated docker](https://github.com/linuxserver/libreelec-builder) (not too sure if it works out with the pine)

**Comment by streppuiu** [comment]
Hi CvH , that is definitely an interresting option. I however am not familiar with docker. If I'm not mistaking I have to import my files to that git repo and then create a docker image on my pine and run it so that it runs as a virtual ubuntu machine inside the pine. I am probably mistaking but I'll look into it and will probably end up using it. Thank you very much for your suggestion CvH.

Any idea why kodi for oe does not compile on pine but throws that error above?

**Comment by stefansaraev** [comment]
cross-compiling OE / kodi on a non x86-64 host was never supported, nor tested. if it works - it's by luck.

**Comment by streppuiu** [comment]
Ok, thank you stefansaraev. I was about to try compiling it on the Rpi itself :) . I just need a small build system to have available so that I get rid of that chest above my frige. I'll probably try the "docker" option. Thank you all for your answers.

**Comment by stefansaraev** [comment]
better dont waste your time with docker on pine64. it's not going to work. get a powerful enough x86_64 hardware. even a cheap $2/mo vps would be less pain.

**Comment by streppuiu** [comment]
Yes, I found the following in the kodi's config.log file on my other pc I built it before:

configure:4572: checking build system type
configure:4586: result: i686-pc-linux-gnu

So, since the software id (linux-gnu) is the same in both descriptors the problem seems to be the hardware descriptor:  i686-pc .

I know docker is platform agnostic so I am curious what would it return at this point during the compilation process. Is there a way to run docker images as if they would be run on a specific hardware? Otherwise I agree that it will not work, it will fail at least when it gets to this point in the compilation process.

Well there are months I don't compile anything so the money would be wasted on the vps on those months, but thank you for your suggestion.

**Comment by lrusak** [comment]
> I know docker is platform agnostic so I am curious what would it return at this point during the compilation process. Is there a way to run docker images as if they would be run on a specific hardware? Otherwise I agree that it will not work, it will fail at least when it gets to this point in the compilation process.

docker is platform agnostic but in the sense of running a container on one computer will be the same as running it on another computer with the same architecture. Docker still requires the contains to match the architecture of the machine it is running on. In your case you would have to run a aarch64 container (or armhf) on your pine64 so it wouldn't make a difference.

PS. the problem is here, https://github.com/xbmc/xbmc/blob/Krypton/m4/xbmc_arch.m4#L35-L74

PPS. I wouldn't try and fix it. as seo said, use a $2 VPS and you won't have any headache.

**Comment by streppuiu** [comment]
Hi lrusak, thank you for the help. I unpacked kodi, changed the file manually and ran the build again and it seems it went through, I am attaching its config.log file below.
[kodi.config.zip](https://github.com/OpenELEC/OpenELEC.tv/files/659599/kodi.config.zip)

I added the following to the end of that switch statement:

  aarch64-*-linux-gnu*)
     AC_SUBST(NATIVE_ARCH_DEFINES, "-DTARGET_POSIX -DTARGET_LINUX -D_LINUX")
     ;;

I will go on with the build, let's see what happens. I have no headache, on the contrary I am doing this for fun. I am in no hurry either, I can try this for weeks :) and most likely with your help I will succeed.

**Comment by streppuiu** [comment]
It seems it completed successfully the entire build. I will try to go on with oe 7.0.0 doing the same fix.
Maybe you could create a patch for kodi to fix this. I know I am the only one interrested in this at this time but who knows, it's a very simple fix. Let me know if you need me to test this on the pine to make sure the images are indeed ok.

When I complete both builds I will try them both on my pi and let you (the ones not interrested :) ) know if it works or I'll ask for help it does not. Thank you.

**Comment by streppuiu** [comment]
It seems to have worked fine. I have the images in my target folder so I assume all ended fine. 
I have create a patch file (almost) manually and used it for kodi (placed it in the packages/mediacenter/kodi/patches folder. I'm not sure it's perfect but here it is, maybe someone is interrested. [kodi-160.01-oe7.0.0_Allow_m4_compilation_on_pine64.zip](https://github.com/OpenELEC/OpenELEC.tv/files/664423/kodi-160.01-oe7.0.0_Allow_m4_compilation_on_pine64.zip)

Thank you lrusak for the help.
